### PR TITLE
Web Callbacks

### PR DESCRIPTION
**Pull request by DavidBoike**
Relates to #753

So far, only V5 sample done.

### CODE REVIEW

**Comment by DavidBoike** [comment]
Now V3-V5 covered. Need V6 and updates to md files to incorporate snippets.

**Comment by DavidBoike** [comment]
@Particular/documentation do you think I should remove some of the cruft that isn't really needed but comes with the MVC template? jQuery, Modernizr, Bootstrap JS (need the CSS), glyphicons?

**Comment by SimonCropp** [comment]
@DavidBoike +1 for removing cruft. but i think something is wrong with this PR. i went to review it and it said  56,363 additions

**Comment by DavidBoike** [comment]
3 new MVC projects, even on the Empty template, will do that. I'll decruftify as much as I can and squash.

**Comment by SimonCropp** [comment]
cool. let me know and we will push it in

File: samples/callbacks/Version_3/WebSender/App_Start/ServiceBus.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        public static IBus Bus { get; private set; }
+
+        static readonly object padlock = new object();
+
+        public static void Init()
+        {
+            if (Bus != null)
+                return;
+
+            lock (padlock)
```
Comment (SimonCropp) [inline samples/callbacks/Version_3/WebSender/App_Start/ServiceBus.cs]: why a lock and not a static constructor?

File: samples/callbacks/Version_3/WebSender/Global.asax.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace WebSender
+{
+    public class MvcApplication : System.Web.HttpApplication
+    {
+        protected void Application_Start()
+        {
+            AreaRegistration.RegisterAllAreas();
+            RouteConfig.RegisterRoutes(RouteTable.Routes);
+            ServiceBus.Init();
```
Comment (SimonCropp) [inline samples/callbacks/Version_3/WebSender/Global.asax.cs]: this wont be needed if we do a static constructor for ServiceBus

File: samples/callbacks/Version_3/WebSender/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+﻿using System.Reflection;
+using System.Runtime.InteropServices;
+
```
Comment (SimonCropp) [inline samples/callbacks/Version_3/WebSender/Properties/AssemblyInfo.cs]: redundant file

File: samples/callbacks/Version_3/WebSender/Web.Debug.config [hunk 4]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<!-- For more information on using web.config transformation visit http://go.microsoft.com/fwlink/?LinkId=125889 -->
+
+<configuration xmlns:xdt="http://schemas.microsoft.com/XML-Document-Transform">
+  <!--
+    In the example below, the "SetAttributes" transform will change the value of 
+    "connectionString" to use "ReleaseSQLServer" only when the "Match" locator
```
Comment (SimonCropp) [inline samples/callbacks/Version_3/WebSender/Web.Debug.config]: remove all the comments. it is just noise

File: samples/callbacks/Version_3/WebSender/Web.Release.config [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<!-- For more information on using web.config transformation visit http://go.microsoft.com/fwlink/?LinkId=125889 -->
+
+<configuration xmlns:xdt="http://schemas.microsoft.com/XML-Document-Transform">
+  <!--
+    In the example below, the "SetAttributes" transform will change the value of 
+    "connectionString" to use "ReleaseSQLServer" only when the "Match" locator 
+    finds an attribute "name" that has a value of "MyDB".
+
```
Comment (SimonCropp) [inline samples/callbacks/Version_3/WebSender/Web.Release.config]: remove comments

File: samples/callbacks/Version_3/WebSender/WebSender.csproj [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+          </CustomServerUrl>
+          <SaveServerSettingsInUserFile>False</SaveServerSettingsInUserFile>
+        </WebProjectProperties>
+      </FlavorProperties>
+    </VisualStudio>
+  </ProjectExtensions>
+  <!-- To modify your build process, add your task inside one of the targets below and uncomment it. 
+       Other similar extension points exist, see Microsoft.Common.targets.
+  <Target Name="BeforeBuild">
+  </Target>
```
Comment (SimonCropp) [inline samples/callbacks/Version_3/WebSender/WebSender.csproj]: remove comments

File: samples/callbacks/Version_3/WebSender/web.config [hunk 7]
```diff
@@ -0,0 +1,61 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!--
+  For more information on how to configure your ASP.NET application, please visit
+  http://go.microsoft.com/fwlink/?LinkId=301880
```
Comment (SimonCropp) [inline samples/callbacks/Version_3/WebSender/web.config]: remove comments

File: samples/callbacks/Version_4/WebSender/App_Start/ServiceBus.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        public static IBus Bus { get; private set; }
+
+        static readonly object padlock = new object();
+
+        public static void Init()
+        {
+            if (Bus != null)
+                return;
+
+            lock (padlock)
```
Comment (SimonCropp) [inline samples/callbacks/Version_4/WebSender/App_Start/ServiceBus.cs]: static constructor

File: samples/callbacks/Version_4/WebSender/App_Start/ServiceBus.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+                Configure.Serialization.Json();
+                Configure configure = Configure.With();
+                configure.Log4Net();
+                configure.DefineEndpointName("Samples.Callbacks.Sender");
+                configure.DefaultBuilder();
+                configure.InMemorySagaPersister();
+                configure.UseInMemoryTimeoutPersister();
+                configure.InMemorySubscriptionStorage();
+                configure.UseTransport<Msmq>();
+                Bus = configure.UnicastBus().CreateBus().Start(() => configure.ForInstallationOn<Windows>().Install());
```
Comment (SimonCropp) [inline samples/callbacks/Version_4/WebSender/App_Start/ServiceBus.cs]: does any cleanup/using need to happen here?

File: samples/callbacks/Version_4/WebSender/Controllers/HomeController.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        #endregion
+
+        #region Web_SendObjectMessage
+        public async Task<ActionResult> SendObjectMessage()
+        {
+            ObjectMessage message = new ObjectMessage();
+            Task<ObjectResponseMessage> responseTask = ServiceBus.Bus.Send("Samples.Callbacks.Receiver", message)
+                .Register<ObjectResponseMessage>(completion =>
+                {
+                    return completion.Messages[0] as ObjectResponseMessage;
```
Comment (SimonCropp) [inline samples/callbacks/Version_4/WebSender/Controllers/HomeController.cs]: dont as when you know the type. do a cast instead

File: samples/callbacks/Version_4/WebSender/Properties/AssemblyInfo.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
+﻿using System.Reflection;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
```
Comment (SimonCropp) [inline samples/callbacks/Version_4/WebSender/Properties/AssemblyInfo.cs]: redundant file

File: samples/callbacks/Version_4/WebSender/Web.Debug.config [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+<!-- For more information on using web.config transformation visit http://go.microsoft.com/fwlink/?LinkId=125889 -->
+
+<configuration xmlns:xdt="http://schemas.microsoft.com/XML-Document-Transform">
+  <!--
+    In the example below, the "SetAttributes" transform will change the value of 
+    "connectionString" to use "ReleaseSQLServer" only when the "Match" locator 
+    finds an attribute "name" that has a value of "MyDB".
+    
+    <connectionStrings>
```
Comment (SimonCropp) [inline samples/callbacks/Version_4/WebSender/Web.Debug.config]: remove comments

File: samples/callbacks/Version_4/WebSender/Web.Release.config [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  <system.web>
+    <compilation xdt:Transform="RemoveAttributes(debug)" />
+    <!--
+      In the example below, the "Replace" transform will replace the entire 
+      <customErrors> section of your web.config file.
+      Note that because there is only one customErrors section under the 
+      <system.web> node, there is no need to use the "xdt:Locator" attribute.
+      
+      <customErrors defaultRedirect="GenericError.htm"
+        mode="RemoteOnly" xdt:Transform="Replace">
```
Comment (SimonCropp) [inline samples/callbacks/Version_4/WebSender/Web.Release.config]: remove comments

File: samples/callbacks/Version_4/WebSender/Web.config [hunk 14]
```diff
@@ -0,0 +1,61 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!--
```
Comment (SimonCropp) [inline samples/callbacks/Version_4/WebSender/Web.config]: remove comments

File: samples/callbacks/Version_6/WebSender/packages.config [hunk 15]
```diff
@@ -0,0 +1,12 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="bootstrap" version="3.0.0" targetFramework="net45" />
+  <package id="jQuery" version="1.10.2" targetFramework="net45" />
+  <package id="Microsoft.AspNet.Mvc" version="5.2.2" targetFramework="net45" />
+  <package id="Microsoft.AspNet.Razor" version="3.2.2" targetFramework="net45" />
+  <package id="Microsoft.AspNet.WebPages" version="3.2.2" targetFramework="net45" />
+  <package id="Microsoft.Web.Infrastructure" version="1.0.0.0" targetFramework="net45" />
+  <package id="Modernizr" version="2.6.2" targetFramework="net45" />
```
Comment (SimonCropp) [inline samples/callbacks/Version_6/WebSender/packages.config]: do we need Modernizr and jQuery and bootstrap?

File: samples/callbacks/Version_6/WebSender/WebSender.csproj [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+      <Name>Shared</Name>
+    </ProjectReference>
+  </ItemGroup>
+  <PropertyGroup>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">10.0</VisualStudioVersion>
+    <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+  </PropertyGroup>
+  <Import Project="$(MSBuildBinPath)\Microsoft.CSharp.targets" />
+  <Import Project="$(VSToolsPath)\WebApplications\Microsoft.WebApplication.targets" Condition="'$(VSToolsPath)' != ''" />
+  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v10.0\WebApplications\Microsoft.WebApplication.targets" Condition="false" />
```
Comment (SimonCropp) [inline samples/callbacks/Version_6/WebSender/WebSender.csproj]: this is a `Condition="false"` and hence never evaluated. u can delete

File: samples/callbacks/Version_6/WebSender/WebSender.csproj [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <Content Include="Views\web.config" />
+    <Content Include="packages.config">
+      <SubType>Designer</SubType>
+    </Content>
+    <Content Include="Views\_ViewStart.cshtml" />
+    <Content Include="Views\Shared\_Layout.cshtml" />
+    <Content Include="Views\Home\Index.cshtml" />
+    <None Include="Web.Debug.config">
```
Comment (SimonCropp) [inline samples/callbacks/Version_6/WebSender/WebSender.csproj]: no point having release and debug config

File: samples/callbacks/Version_3/WebSender/App_Start/ServiceBus.cs (hunk 1 above)
Comment (DavidBoike) [inline samples/callbacks/Version_3/WebSender/App_Start/ServiceBus.cs]: I've always used this pattern in web applications. Requiring an Init() referenced from the Global.asax, to me, follows principle of least surprise - you can look at Global.asax and know all the things that are happening, and get to choose when it happens. Static constructor is decidedly less deterministic.

File: samples/callbacks/Version_4/WebSender/App_Start/ServiceBus.cs (hunk 9 above)
Comment (DavidBoike) [inline samples/callbacks/Version_4/WebSender/App_Start/ServiceBus.cs]: Do you think it does? Bus instance is going to live until the AppDomain shuts down.

**Comment by DavidBoike** [comment]
Finished this up. Removing WIP flag. @Particular/documentation please review.

**Comment by SimonCropp** [comment]
can u squash this so we dont have the history of 70000 additions and 69000 removals https://github.com/Particular/docs.particular.net/commit/b9b6d91ea45c86d9e149f9797024c8f8662342c0

**Comment by DavidBoike** [comment]
@SimonCropp Squash and rebase complete

**Comment by SimonCropp** [comment]
the solutions are referencing inconsistent nugets

eg this https://github.com/Particular/docs.particular.net/blob/web-callbacks/samples/callbacks/Version_6/WebSender/packages.config has 
```
  <package id="NServiceBus" version="6.0.0-unstable0601" targetFramework="net451" />
  <package id="NServiceBus.Callbacks" version="1.0.0-unstable0011" targetFramework="net451" />
```

and this https://github.com/Particular/docs.particular.net/blob/web-callbacks/samples/callbacks/Version_6/Sender/packages.config has
```
  <package id="NServiceBus" version="6.0.0-unstable0815" targetFramework="net451" />
  <package id="NServiceBus.Callbacks" version="1.0.0-unstable0017" targetFramework="net451" />
```

seems to be similar problems in the other versions as well

**Comment by SimonCropp** [comment]
packages are reference in wrong level

eg this https://github.com/Particular/docs.particular.net/blob/web-callbacks/samples/callbacks/Version_6/Sender/Sender.csproj

uses 
```
<HintPath>..\..\..\..\packages\xxxl</HintPath>
```

but this https://github.com/Particular/docs.particular.net/blob/web-callbacks/samples/callbacks/Version_6/WebSender/WebSender.csproj#L70
```
 <HintPath>..\packages\xxx</HintPath>
```

**Comment by SimonCropp** [comment]
remove the redundant `Model` and `Properties` directories

**Comment by SimonCropp** [comment]
this will need to be 4.5.2 https://github.com/Particular/docs.particular.net/blob/web-callbacks/samples/callbacks/Version_6/WebSender/WebSender.csproj#L16

**Comment by DavidBoike** [comment]
@SimonCropp fixed all those things. Please take another look.

**Pull request merged**